projects/storefrontlib/shared/components/captcha/captcha.service.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * SPDX-FileCopyrightText: 2023 SAP Spartacus team <spartacus-team@sap.com>
+ * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Injectable, OnDestroy } from '@angular/core';
+import {
+  BaseSiteService,
+  CaptchaConfig,
+  LanguageService,
+  ScriptLoader,
+  SiteAdapter,
+} from '@spartacus/core';
+import { forkJoin, map, Observable, ReplaySubject, Subscription } from 'rxjs';
+import { concatMap, take } from 'rxjs/operators';
+import { CaptchaProvider, RenderParams } from './captcha.model';
+import { CaptchaApiConfig } from './captcha-api-config';
+
+/**
+ * Global function to be passes as "onload" url param for captcha <script>, to be
+ * triggered once script and dependencies are loaded
+ */
+declare global {
+  interface Window {
+    onCaptchaLoad: () => void;
+  }
+}
+
+@Injectable({
+  providedIn: 'root',
+})
+export abstract class CaptchaService implements CaptchaProvider, OnDestroy {
```

Comment:
I find the break-down into `CaptchaProvider`, `CaptchaService`, and `CaptchaComponent` extremely confusing.

Shouldn't a service be reusable by multiple components? This `CaptchaService` holds `CaptchaProvider` model data that could get be associated with multiple instances of `CaptchaComponent`. Customers might also inject this `CaptchaService` into other components altogether.

Depending on the implementation of `CaptchaService`, the same `token` property could get set twice with differnt values. I guess this might work in the end, since we don't care about a specific `token`, just any `token` that the backend (OCC API) can verify against the captcha service successfully. Still this whole design looks very weird to me.

Reply:
Yep, they are complex and confusing. 
CaptchaComponent : Angular component which could be inserted in different places
CaptchaService: Abstract class which provide fetch settings from backend and other common features
CaptchaProvider: Which contains configuration of how it rendering, part of CaptchaApiConfig


> Depending on the implementation of CaptchaService, the same token property could get set twice with differnt values. I guess this might work in the end, since we don't care about a specific token, just any token that the backend (OCC API) can verify against the captcha service successfully. Still this whole design looks very weird to me.

Yep,  same/different tokens may set many times based on how/whether the rendering is called and fetch token from server side, we always communicate with Commerce backend with latest one.